### PR TITLE
Feat/implement parsing of gui commands

### PR DESCRIPTION
**Pull request by Njord201**
I've added the parsing of gui comands, and the response to all commands. I also update the server structure for fit to the new commands.

### CODE REVIEW

File: server/src/gui/communication/pin.c [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+#include "rules.h"
+#include "ai/ai.h"
+#include "app/app.h"
+#include "gui/gui.h"
+#include "parsing.h"
+
+#include <stddef.h>
+#include <stdbool.h>
+
+static bool is_str_number(char *str)
```
Comment (mathieurobert1) [inline server/src/gui/communication/pin.c:17]: you repeat this function in many file, please make an util file.

**Review by mathieurobert1 — CHANGES_REQUESTED** [review]
A unique request

File: server/src/gui/communication/pgt.c [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+#include "ai/ai.h"
+#include "app/app.h"
+
+void pgt_command(app_t *app, int player_id)
+{
+    char *response = NULL;
+    list_node_t *gui_node = app->gui_list->first;
+
+    for (int i = 0; i < 7; i++) {
```
Comment (theophile-jr) [inline server/src/gui/communication/pgt.c:16]: MACRO

File: server/src/gui/communication/ppo.c [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        ia->fd, ia->position->x, ia->position->y, ia->direction);
+    return response;
+}
+
+void ppo_response(gui_t *gui, app_t *app, char *line)
+{
+    int player_id = 0;
+    ia_t *ia = NULL;
+
+    if (line[3] != ' ') {
```
Comment (theophile-jr) [inline server/src/gui/communication/ppo.c:33]: MACRO // same for other files.

File: server/src/gui/communication/sst.c [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+void sst_response(gui_t *gui, app_t *app, char *line)
+{
+    char *response = NULL;
+    int freq = 0;
+
+    if (line[3] != ' ') {
+        suc_command(gui);
+        return;
+    }
+    line += 4;
```
Comment (theophile-jr) [inline server/src/gui/communication/sst.c:29]: //

**Review by theophile-jr — CHANGES_REQUESTED** [review]
request changes + talked w njord IRL about some issues.

File: server/include/gui/communication.h [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+** EPITECH PROJECT, 2024
+** Zappy
+** File description:
+** Header file for the communication functions between the GUI and the server.
+*/
+
+#pragma once
+
+#include "../app/app.h"
+#include "gui.h"
```
Comment (Thomaltarix) [inline server/include/gui/communication.h:11]: Wrong include order

File: server/src/gui/communication/sst.c [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+#include "app/app.h"
+#include "gui/gui.h"
+#include "parsing.h"
+#include "gui/communication.h"
+
+#include <stddef.h>
+#include <stdbool.h>
+
+
+// TODO error
```
Comment (Thomaltarix) [inline server/src/gui/communication/sst.c:19]: I think you can precise what's needed

File: server/src/app/app.c [hunk 7]
```diff
@@ -27,6 +27,7 @@ app_t *create_app(parsing_t *parsing)
     new_app->gui_list = malloc(sizeof(list_t));
     new_app->clients_list = malloc(sizeof(list_t));
     new_app->teams_list = malloc(sizeof(list_t));
+    new_app->eggs_list = malloc(sizeof(list_t));
     create_teams(parsing, new_app);
     if (new_app->gui_list == NULL || new_app->clients_list == NULL
     || new_app->teams_list == NULL)
```
Comment (Thomaltarix) [inline server/src/app/app.c:33]: You're adding a malloc but not adding its verification ?

File: server/src/gui/communication/loop.c [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    [CL_BCT] = &bct_response,
+    [CL_MCT] = &mct_response,
+    [CL_TNA] = &tna_response,
+    [CL_PPO] = &ppo_response,
+    [CL_PLV] = &plv_response,
+    [CL_PIN] = &pin_response,
+    [CL_SGT] = &sgt_response,
+    [CL_SST] = &sst_response,
+    [CL_LEN] = NULL
+};
```
Comment (Thomaltarix) [inline server/src/gui/communication/loop.c:47]: For me, since its something constant, that doesn't change, it should be in a header file

**Review by Thomaltarix — CHANGES_REQUESTED** [review]
Some changes to request, but overall great

File: server/src/app/app.c [hunk 9]
```diff
@@ -27,6 +27,7 @@ app_t *create_app(parsing_t *parsing)
     new_app->gui_list = malloc(sizeof(list_t));
     new_app->clients_list = malloc(sizeof(list_t));
     new_app->teams_list = malloc(sizeof(list_t));
+    new_app->eggs_list = malloc(sizeof(list_t));
```
Comment (mathieurobert1) [inline server/src/app/app.c:30]: check your malloc

**Review by mathieurobert1 — CHANGES_REQUESTED** [review]
Some requests.

**Review by mathieurobert1 — DISMISSED** [review]
Seems good, hope @AxelF44 tests will be approved.

File: server/src/server/read/read_data.c [hunk 10]
```diff
@@ -66,7 +67,9 @@ void handle_request(app_t *app, size_t fd, char *line)
     ia_t *ai = find_ia(app, fd);
 
     if (gui != NULL) {
-        (void) line;
+        if (gui->list_messages->len >= 10)
+            return;
```
Comment (AxelF44) [inline server/src/server/read/read_data.c:71]: Remove this it's useless.

**Review by AxelF44 — CHANGES_REQUESTED** [review]

File: server/src/ai/ai.c [hunk 11]
```diff
@@ -78,41 +79,42 @@ static time_info_t *create_time(app_t *app)
...
-        team->egg_position->len > 0) {
-            data.ai = create_ia(app, fd, team);
-            list_add_back(team->list_ai, data);
+        if (strcmp(line, temp->data.team->name) == 0 &&
+        temp->data.team->eggs_list->len > 0) {
+            data.ai = create_ia(app, fd, temp->data.team);
+            list_add_back(temp->data.team->list_ai, data);
             free(client_node->data.client);
             list_delete(app->clients_list, client_node);
+            free(temp->data.team->eggs_list->first->data.egg);
```
Comment (AxelF44) [inline server/src/ai/ai.c:116]: You free the egg struct but not the struct vector_i !

**Review by AxelF44 — CHANGES_REQUESTED** [review]

**Review by theophile-jr — APPROVED** [review]

**Review by mathieurobert1 — APPROVED** [review]
Seems good

**Review by AxelF44 — APPROVED** [review]

**Review by mathieurobert1 — APPROVED** [review]
Nice :+1:

**Review by theophile-jr — APPROVED** [review]

**Review by Thomaltarix — APPROVED** [review]

**Pull request merged**